segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java

```diff
@@ -1452,7 +1453,7 @@ protected void declareVersions() {
             }
 
             private void read00(RevisionDataInput input, SystemJournalRecordBatchBuilder b) throws IOException {
-                b.systemJournalRecords(input.readCollection(ELEMENT_DESERIALIZER));
+                b.systemJournalRecords(input.readCollection(ELEMENT_DESERIALIZER, () -> new Vector<>()));
```

Comment:
Why is this needed? 
Is the `systemJournalRecords` member variable mutated after the creation of the object?

Reply:
Here the issue is **not** concurrent modification - it is the issue of thread visibility. (processor cache flush.  updates made by one thread are not visible to other thread. [java memory model](https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html))

ArrayList created by default is non thread safe data structure. There is no guarantee that changes made by one thread is immediately visible to other thread. 


Advice given here is cumbersome to implement with lots of async code.
https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#synchronizedList-java.util.List-

>   It is imperative that the user manually synchronize on the returned list when iterating over it:
>   List list = Collections.synchronizedList(new ArrayList());
>       ...
>   synchronized (list) {
>       Iterator i = list.iterator(); // Must be in synchronized block
>       while (i.hasNext())
>           foo(i.next());
>   }

That is why given the context and usage (this code get used very rarely and using Vector is not an issue.)

Reply:
This doesn't apply if this is being used correctly. Which is why I am concerned. If this is suppressing a bug, I fear it is not actually fixed. 
In this case an object is being build with a builder and then it's constructor is called inside the `.build()` which assigned to SystemJournalRecordBatch which is (and should be) a final field. If this list is then never changed (Which is the property I want to make sure holds) then SystemJournalRecordBatch is immutable and no synchronization is required because the construction of everything inside of it has a happens-before relationship to the object itself.
If on the other hand is *not* immutable, then there may be a race between items being added and removed from the list, and the list being read.
If we are following the rule: All data is immutable. Then validating this is trivial. The simplest way to make this clear would to have the member be an `ImmutableList` and use `ImmutableList.builder()` in the call to `readCollection` here.

Reply:
No systemJournalRecords field is not updated after creation.

![image](https://user-images.githubusercontent.com/44757683/165649064-55df7810-7afa-4127-9143-932018e67b9d.png)

Reply:
Definitely not modifying deserialized object.

Reply:
Ok. Great. Then let's use: `ImmutableList` and use `ImmutableList.builder()` it will make this explicit and have higher performance